### PR TITLE
fix rendering types for React 16

### PR DESCRIPTION
**Pull request by Pajn**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/20097
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Apart from imposing artificial limitations that React does need, the current types are quite a pain as none of the named types line up with the result of rendering a component which creates the need to either cast, type with `any` or type with a really long and messy union.

### CODE REVIEW

**Comment by tkrotoff** [comment]
👍  works with my code and solves https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19363#issuecomment-335089941

Would be nice to add tests, example:
```TypeScript
  render() {
    const { children } = this.props;
    return children !== undefined ? children : null;
  }
```

Also SFC still does not work, example:

```TypeScript
interface SCProps {
  foo: string;
  bar: string;
}

const StatelessComponent: React.SFC<SCProps> = ({foo, bar}) => {
  return [
    <div key="1">{foo}</div>,
    <div key="2">{bar}</div>
  ];
};

const sfc = <StatelessComponent foo="foo" bar="bar">SFC</StatelessComponent>;
```

**Comment by Pajn** [comment]
Good spots!
I have added your examples as tests and made sure that they work.

**Comment by RyanCavanaugh** [comment]
Approved by a listed owner. PR ready to merge pending express review by a maintainer.

**Comment by tkrotoff** [comment]
@Pajn 
> Edit:  forget about the SFC part, see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19363#issuecomment-337161610

**Comment by Pajn** [comment]
That can be fixed by fixing JSX.Element, but that seems to break lots of other types. I have omitted the SFC changes now so we should be back to passing tests.

**Comment by Pajn** [comment]
As (apparently) libraries written against the current typings will depend on false facts and break when SFCs are fixed it should be fixed ASAP though. The problem will only grow for every single type written against these typings in the meantime.

**Review by DovydasNavickas — APPROVED** [review]

**Comment by DovydasNavickas** [comment]
Let's merge this now as it helps already and work on the SFC stuff ASAP, as you said.
Can you make another PR with SFC changes or should I?

**Comment by Pajn** [comment]
I pushed the SFC changes to a separate branch, it will need a new commit message and of course many packages are broken https://github.com/Pajn/DefinitelyTyped/commit/e05c3b3bcdd9f0d54e92a61d0005706ed7d0e1ea
On problem I realized with setting `type Element = React.ReactNode;` is that `undefined` is not okay as a sole return from SFC so we might need a new type that is `ReactNode` except for `undefined` that we set `Element` to be. I can't work on it right now so if you wont to do that, go a head. Base your changes on mine if you want.

**Comment by DovydasNavickas** [comment]
@RyanCavanaugh could you merge this one?

**Comment by Jakst** [comment]
Which npm version will this fix be in?

Still not working with these versions:
@types/react: 16.0.22
typescript: 2.6.1

**Comment by GarrisonD** [comment]
Hello guys!

I am using `@types/react@^16.0.34` and it is still not working.

Any updates on it?

**Comment by jshado1** [comment]
@Pajn was this ever fixed for stateless components?

I'm on @types/react 16.3.10 and still getting this error.

**Comment by Pajn** [comment]
I never made a second PR as there were some discussion in another PR (i think, might have been an issue) about it. There are some problems that I don't think can be solved without changes in Typescript. There is an open issue there but I don't have the number available ATM.

**Comment by jshado1** [comment]
@Pajn do you recall the number?

**Comment by eps1lon** [comment]
@jshado1 I believe Microsoft/TypeScript#21699 is the one that would solve those issues.

**Pull request merged**
